src/pybind/rados/rados.pyx

```
@@ -2272,6 +2282,148 @@ class ReadOpCtx(ReadOp, OpCtx):
+            uint64_t _cookie;
+            uint32_t _timeout = timeout;
+            void *_args = <PyObject*>self
+
+        with nogil:
+            ret = rados_watch3(self.ioctx.io, _oid, &_cookie,
+                               <rados_watchcb2_t>&__watch_callback,
+                               <rados_watcherrcb_t>&__watch_error_callback,
+                               _timeout, _args)
+        if ret < 0:
+            raise make_ex(ret, "watch error")
+
+        self.id = int(_cookie);
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type_, value, traceback):
+        self.close()
+        return False
+
+    def __dealloc__(self):
+        self.close()
+
+    def _callback(self, notify_id, notifier_id, watch_id, data):
+        replay = self.callback(notify_id, notifier_id, watch_id, data)
+
+        cdef:
+            rados_ioctx_t _io = <rados_ioctx_t>self.ioctx.io
+            char *_obj = self.oid
+            int64_t _notify_id = notify_id
+            uint64_t _cookie = watch_id
+            char *_replay = NULL
+            int _replay_len = 0
+
+        if replay is not None:
+            replay = cstr(replay, 'replay')
+            _replay = replay
+            _replaylen = len(replay)
+
+        with nogil:
+            rados_notify_ack(_io, _obj, _notify_id, _cookie, _replay,
+                             _replaylen)
+
+    def _error_callback(self, watch_id, error):
+        if self.error_callback is None:
+            return
+        self.error_callback(watch_id, error)
+
+    def get_id(self):
+        return self.id
+
+    def check(self):
+        """
+        Check on watch validity.
+
+        :raises: :class:`Error`
+        :returns: ms since last confirmed valid, or error
+        """
+        self.ioctx.require_ioctx_open()
+
+        cdef:
+            uint64_t _cookie = self.id
+
+        with nogil:
+            ret = rados_watch_check(self.ioctx.io, _cookie)
+
+        return ret
```

Nit: feel free to ignore, but perhaps just return a `datetime.timedelta` to avoid the need to worry about the units.

---

And what about the case when `ret` is a negative error code?

---

Probably should just raise an exception to keep it consistent w/ the other methods

---

Just thinking about a use case. Actually I was thinking the main purpose of the watch check would be to check if the watcher was not in a error state to re-watch if it was. E.g. I could use this in the mirror snapshot scheduler instead of registering the error callback.

But I am fine with your approach and can do this.

---

I was just saying we should never return an error code from these methods -- we should raise an exception if an error occurs since that's what the other pybind methods do.

---

I think I understand you. But when I first saw `rados_watch_check` function I had an impression its main purpose was to check if the watch was not in an error state, and just as a gratis it would also return some useful info when it was not an error. The same as with `read` -- the caller is usually interested only if it returns success.

But it was just mine impression and now I think it was wrong. I am going to make it as you propose -- it makes sense to me.